### ISSUE TITLE
Notebook API: Notification of what NotebookController is selected when document is initially opened

### ISSUE THREAD

**Issue by IanMatthewHuff** [microsoft/vscode#122415]
Opening a new issue per the discussion here:
https://github.com/microsoft/vscode-jupyter/pull/5635#discussion_r621409778

I want to be open about how this could be implemented, so I'll try to stick to just detailing what we'd like. Right now we're only getting onDidChangeNotebookAssociation for controllers specifically on user actions for picking a Controller from the drop down. We'd like to get a notification about when a Controller is initially selected at document open, or if it's changed due to updateNotebookAffinity. Right now when a document is initially opened (or a new blank notebook is created) we don't actually know what NotebookController is associated with this. We'd like to know this so that (if it's one of our controllers) we can start prewarming kernels as well as for our test infrastructure to make sure that the correct initial kernel is selected.

**Comment by IanMatthewHuff** [comment]
@DonJayamanne and @jrieken Opened a new issue to move discussion out from that PR thread.

**Comment by jrieken** [comment]
I feel like you are making fun of me... As stated repeatedly (https://github.com/microsoft/vscode-jupyter/pull/5635#discussion_r621365125)

> Notebook controller selection (what causes the onDidChangeNotebookAssociation-event) is a user action. Extensions can use the affinity API to become more prominent in UI for that user action. Users can explicitly select a controller via the picker or implicitly when running with a suggested controller.

So, what is your suggestion? How should this work?

**Comment by DonJayamanne** [comment]
> So, what is your suggestion? How should this work?

These events are not exposed to all extension authors, only the extension that created the controller. Hence I don't see the need to be strict about raising the event only when a user selected this. It's not an event for other extension authors.

One Solution - fire the event and pass a flag that indicates whether user selected this vs vscode auto selected this.

Further to ian's reasons, we also need this to setup the communication channel with ipywidgets. Without this I'll need to initialize the comms with all controllers - which is wrong.

**Comment by jrieken** [comment]
> Hence I don't see the need to be strict about raising the event only when a user selected this. It's not an event for other extension authors.

Well, it's the opposite. Another extension can appear and always be faster in creating a preferred controller. Then yours is never selected and I don't believe you want that. 

I simply don't see the ability to auto-select (not suggest) a controller and be fair to extensions. So, I still believe the logic of explicitly or implicitly selecting a controller is the only viable approach. Today, implicit selection only happens when users execute cells. Because some kernels also offer language services we should think about more of these implicit selector event, like completion or hover requests.

**Comment by DonJayamanne** [comment]
> So, I still believe the logic of explicitly or implicitly selecting a controller is the only viable approach.

This would be a regression for our extension & not how Jupyter works either.

**Comment by rebornix** [comment]
Since I was testing the controller API, I read through the discussions in the PR and here and try to understand.

@DonJayamanne correct me if I was wrong. What you was asking for is Know if your preferred kernel will be the active kernel when users press run button (and assuming there is no other preferred kernels pushed from any extensions) and then the Jupyter extension can do a warm up for the kernel (even if the users never hit the run button).

**Comment by rebornix** [comment]
> Further to ian's reasons, we also need this to setup the communication channel with ipywidgets. Without this I'll need to initialize the comms with all controllers - which is wrong.


@DonJayamanne can you also elaborate more on this? how do you initialize the comms without the controller being the real active controller?

**Comment by jrieken** [comment]
>  how do you initialize the comms without the controller being the real active controller?

@DonJayamanne Esp, how do you communicate with a renderer which preloads haven't been loaded?

**Comment by IanMatthewHuff** [comment]
> I feel like you are making fun of me...

Very sorry if you are feeling that way, not my intention at all. I was just trying to move the discussion out of the PR and into a new area, and to try to define exactly what we are looking to accomplish. I feel like we're not fully connecting on what we are asking for so I was trying to define it better. There are also some terms that I felt were unclear between the two teams.

> Today, implicit selection only happens when users execute cells.

This the part that is still confusing to me. When I open up a document now there is a kernel selected in the lower right kernel before the document runs. And when we add our preferredKernel this selection changes. I don't understand why this is not being counted as "selection" as being discussed above. To my understanding vscode internally has some concept of what kernel would be run if the user would hit the run button. This kernel that would be run might change as we (or other extensions) do things like register kernels or change preferred kernels.

> I simply don't see the ability to auto-select (not suggest) a controller and be fair to extensions.

So auto-selection is not what I'm trying to ask for here. I know that VS Code's API needs to be fair to all extensions. I just want to know if any of my controllers is the controller that VS Code would current use if the user was to execute cells on the given document. If foo.ipynb is opened I'd like to suggest kernel A (I understand that I'm not in full control of selection) and be notified if VS Code does prefer it (by which I mean it's the current kernel showing in the lower right).

**Comment by DonJayamanne** [comment]
Ok, I think we can solve two issues we are having:
* Ipywidgets - we can setup the comms for preferred controller 
* prewarming - @IanMatthewHuff we can prewarm the preferred kernel (even if we're don't get an event from vscode).

@IanMatthewHuff I think the only other place we're blocked is our tests, we don't know what has been selected/preferred. I guess we can change the tests to call method on the controller manager to check whether the right controller had been setup as preferred. 

Sorry I didn't read the previous messages. Will do so

**Comment by jrieken** [comment]
> Very sorry if you are feeling that way, not my intention at all. I was just trying to move the discussion out of the PR and into a new area, and to try to define exactly what we are looking to accomplish. 

No worries, I like having the discussion here. I reacted because I feel like I have explained selected vs suggested controllers too many times already.

> This the part that is still confusing to me. When I open up a document now there is a kernel selected in the lower right combobox before the document runs. And when we add our preferredKernel this selection changes.

There is two concepts: 

1. A "list of potential controllers" - this is all currently available notebook controllers _filtered_ by viewType and _sorted_ by affinity. 
2. A "selected controller" for a notebook - which is always one from the list of potential controllers and which has been picked through a user gesture. 

There are subtle differences in the UI and I am trying to explain them in the following screen shots (I have dummy notebooks and three available controllers):


**Case 1: Selected controller**

For the `test.lnb`-file the user has selected a controller ("Rundam"). In the picker that is shown with the "Currently Selected" suffix.

<img width="770" alt="Screenshot 2021-04-28 at 19 15 02" src="https://user-images.githubusercontent.com/1794099/116449747-c4f75a80-a85a-11eb-898d-637118312558.png">

**Case 2: No selected controller**

For the `test-2.lnb`-file _no_ controller has been selected. In the picker no controller shows as selected and tooltip in the status bar has the little "suggestion"-suffix

<img width="795" alt="Screenshot 2021-04-28 at 19 15 55" src="https://user-images.githubusercontent.com/1794099/116449973-0b4cb980-a85b-11eb-9ad7-fc07c34113d6.png">

So, in case 2 there is a list of controllers and none of them has been selected. Despite the similar UI for selected and suggested controllers, it's up to the user to select one. This selection happens explicitly from the picker or implicitly when running a cell while a controller has been _suggested_. 

I am happy to tweak the UI, like add more reasons to implicitly select a controller or make implicit selection more explicit. However, the list of potential controllers and a _user selected_ controllers is here to stay. Extensions control the list of potential controllers and users select from that list. It's not extensions selecting from that list.


>  I just want to know if any of my controllers is the controller that VS Code would currently use if the user was to execute cells on the given document

This is the catch. The UI only shows one _suggested_ controller but in fact there can be many preferred controllers for a single notebook. So, we would have a bucket of controllers that are likely to be selected and we would need to inform all of them. But that only means that we are telling you something that you already knew - you already knew that you are preferred because you told us so.

To summarize: 

* I want us to agree on the difference between potential controllers and user-selected controllers and that only users pick controllers.
* We all agree that the UI is bad and needs improvement. We could do away with the implicit stuff and be more explicit, e.g. show picker on run, show stronger status bar messages etc
* Concrete feedback please! Saying A, B, C and D doesn't work without further depth is not helpful

**Comment by rebornix** [comment]
> I want us to agree on the difference between potential controllers and user-selected controllers and that only users pick controllers.

@jrieken one misunderstanding I found during discussion with @IanMatthewHuff (you can correct me if I was wrong) is the status bar item is not telling you if the controller is selected or suggested, though it's clear in either the quick pick or the API events.

**Comment by jrieken** [comment]
> @jrieken one catch I found during discussion with @IanMatthewHuff (you can correct me if I was wrong) is the status bar item is not telling you if the controller is selected or suggested,

Yeah, only through the status bar tooltip. Not the best UI...

**Comment by IanMatthewHuff** [comment]
@rebornix and I just had a long chat on teams regarding this in the AM.

To be clear, what @DonJayamanne mentioned here I've already done. 

> @IanMatthewHuff I think the only other place we're blocked is our tests, we don't know what has been selected/preferred. I guess we can change the tests to call method on the controller manager to check whether the right controller had been setup as preferred.

When we set the preferred controller on document open at the same time I send basically a fake "selected" method. For the purposes of our code when we set a preferred controller we consider that controller selected and start our prewarming and whatnot. Also for our tests we can then tell what the preferred controller is. What we have now I believe we can work with. @DonJayamanne do you agree? Not fully sure if there are widget issues that I'm not aware of currently. 

Per @jrieken 's post.
> I want us to agree on the difference between potential controllers and user-selected controllers and that only users pick controllers.

Yup, we are in agreement on this. The confusion for me on this was more around the state that preferred controllers were in before user-selected controllers were added into the mix. Relevant to this comment here:

> This is the catch. The UI only shows one suggested controller but in fact there can be many preferred controllers for a single notebook.

There were two confusing bits about the status bar UI regarding this for me.

- From the status bar UI in the state where no controller is selected there appears to be **a** preferred controller, or at least a more preferred controller, not a bucket of preferred controllers. Per your comments above this is not the case. I was confusing the one suggested controller with a more highly preferred controller.
- If you execute in this state where there is no user selected controller yet the implicit selection happens at execution time. Given that one controller was explicitly shown in the status bar UI my assumption was that this implicit selection was basically happening in real time, continually updating as more controllers were added or if new affinities were set. This is not the case though, until execution there is just a bucket of preferred items, and no implicit selection happens until execution.

A few things that I was discussing with @rebornix from this AM.

- I was concerned about knowing when and how updateNotebookAffinity would take effect. In our test code if I know that I've called updateNotebookAffinity with our preferred controller do I know that right away I can execute a cell in the file and get that preferred controller via the implicit selection? From the discussion above I believe that I can rely on this. Since updateNotebookAffinity is sync as long as it has returned I should be safe to execute as my preferred controller would be added to the preferred controllers bucket and would be selected by the selection that takes place during first execution. 
- If our code to find a preferred controller fails (so a document opens, but we don't call updateNotebookAffinity for any of our NotebookControllers on it) what controller would we prewarm? Given the discussion above the UI would just be showing a list of all the registered controllers and would pick one of them on execution. I think this just means that in our extension we need to make sure that if we want to prewarm we need to pick a preferred controller. If we don't have metadata in the file, just fall back on some heuristic for the most reasonable controller that we have.
- If we optimistically pre-warm when we set a preferred controller what about the case where another extension also registers a preferred controller. As I understand, at execution time vscode will just select between the two controllers based on some heuristic and we might have prewarmed a controller that doesn't actually execute when the user runs. I think this could be possible, but at this point it just is what it is. Our extension will just prewarm and not be called on execute. Probably not a huge deal.

Summary: I think that we have a better understanding on our side now. For the Jupyter extension we can get the behavior that we want by doing what we are already doing, on document open setting a single preferred controller and then treating that controller as the expected target launch controller (prewarming and what not) in our code. In our test code we can assume that as soon as we set controller affinity we can execute on the document and vs code will have that preference.

**Comment by jrieken** [comment]
>  For the Jupyter extension we can get the behavior that we want by doing what we are already doing, on document open setting a single preferred controller and then treating that controller as the expected target launch controller (prewarming and what not) in our code.

That is good to hear and also sounds reasonable to me. In my (simplified?) mental model jupyter does the following

1. When being activated find jupyter-kernels and create corresponding notebook controllers
2. When notebook documents (ipynb) are being opened set affinity and optimistically pre-warm kernel etc
3. When a pre-warmed kernel is not used, then tear it down after some time has passed


For the UI side, I am working on making things more explicit. There is two cases for which I have a proposal, both are for the case in which no controller has been selected yet

**Case 1**: no kernel has been selected, there is only 1 kernel available

This is the github issue notebook case and presumably the case for many "simple" notebooks that only have a single controller. We stick to the concept of implicit selection. Today, only execute is an implicit selection trigger but we are thinking of adding more of these implicit trigger reasons. 

https://user-images.githubusercontent.com/1794099/116534552-f3b91380-a8e2-11eb-902d-98f9e40997e9.mov

_There is only one controller for the 'test.nodebook'-file. That controller is suggested and becomes selected when pressing run. 🤔 The status bar picker is disabled but we could also show a picker with a single entry_

**Case 2**: no kernel has been selected, multiple kernels are available

This is the jupyter case and the UI is now making you explicitly select a controller. The status bar entry is more prominent and shows a selection affordance. Also, when selecting run you must select a controller to continue

https://user-images.githubusercontent.com/1794099/116535012-8bb6fd00-a8e3-11eb-8245-a5c84c9b22fb.mov

_There are multiple controllers for the `test5.lnb`-file. The status bar entry shows more prominent and wants you to make a selection. Pressing run also shows the picker first. Note the slight difference in the picker: its placeholder says "select controller" vs "change controller"_

**Comment by jrieken** [comment]
> 🤔 The status bar picker is disabled but we could also show a picker with a single entry

Replying to myself: ✔️ did that, even with a single controller you can pick (and select) that controller

**Comment by kieferrm** [comment]
The one issue I see here is that the controller terminology which works well on the API side now flows through to the Jupyter world. There, as a user, I don't know what a controller is, but I do know kernels. We should use the kernel terminology in the UI.

**Comment by DonJayamanne** [comment]
Thanks @jrieken 
I believe https://github.com/microsoft/vscode/issues/122415#issuecomment-829109276
Will resolve the issues we have, and the behavior will be similar to what it was prior to kernel push

**Comment by jrieken** [comment]
Thanks for all the feedback. Closing since the big chunk of work is done, let's track the rest in individual issues
